content/faq/serverless.md

```diff
@@ -209,36 +223,53 @@ functions:
 
 > info **Hint** To learn more about the Serverless framework, visit the [official documentation](https://www.serverless.com/framework/docs/).
 
-With this place, we can now navigate to the `main.ts` file and update our bootstrap code with the required boilerplate:
+With that in place, we can now create a file called `lambda.ts` inside `src` and place the necessary boilerplate:
 
 ```typescript
 import { NestFactory } from '@nestjs/core';
-import serverlessExpress from '@vendia/serverless-express';
-import { Callback, Context, Handler } from 'aws-lambda';
+import { default as serverlessExpress } from '@vendia/serverless-express';
+import type { Callback, Context, Handler } from 'aws-lambda';
 import { AppModule } from './app.module';
 
 let server: Handler;
 
 async function bootstrap(): Promise<Handler> {
   const app = await NestFactory.create(AppModule);
+
   await app.init();
 
   const expressApp = app.getHttpAdapter().getInstance();
+
   return serverlessExpress({ app: expressApp });
 }
 
+// This is a simple trick inspired by this article
+// https://aws.amazon.com/pt/blogs/compute/using-node-js-es-modules-and-top-level-await-in-aws-lambda/
+// This can save you up to one second from cold start.
+let delayedFactory: Promise<Handler> = Promise.resolve()
+  .then(() => bootstrap())
+  .then(app => {
+    server = app;
+
+    return app;
+  });
+
 export const handler: Handler = async (
   event: any,
   context: Context,
   callback: Callback,
 ) => {
-  server = server ?? (await bootstrap());
+  server = server ?? (await delayedFactory);
+
   return server(event, context, callback);
 };
+
 ```
 
 > info **Hint** For creating multiple serverless functions and sharing common modules between them, we recommend using the [CLI Monorepo mode](/cli/monorepo#monorepo-mode).
 
+> info **Hint** It's really recommended to have one file called `setup.ts` to share common configurations to use inside `lambda.ts` and `main.ts`.
```

Comment:
As someone who doesn't work with serverless, is it common to have a `main.ts` for controlled server applications and `lambda.ts` for serverless, or would you normally pick one or the other? Really wondering what the use for both bootstrap files is

Reply:
Imagine that you need to configure middlewares, cors, swagger, filters, configure global validation pipes, etc. In my company we use a file to put all this configuration and export as `createApp()`.

In `main.ts` we just run `app.listen()` and inside `lambda.ts` we don't need to call `listen` and we just need to export the handler to AWS.

I didn't put in the documentation the creation of this `setup.ts` file to simplify the process but I can put it instead of leaving a hint.

Reply:
Just to give some example, see [this method](https://github.com/H4ad/bird-app/blob/master/backend/src/main.ts#L70), a better organization is to move the `setup` function to another file and then import inside `main.ts`.

Reply:
I understand the use of the `setup` method, I use it for running my application and e2e tests in a similar manner. Just not aware of how `serverless offline` acts and thus why there is a need for a `lambda.ts` and a `main.ts`.

Reply:
When you run `serverless offline`, you need to export a function called `handler`. Next, the library will create an http server, then proxy browser requests to the `handler`, then re-proxy the NestJS application with `serverless-express`.

Theoretically, you don't need both, you have both for convenience:

- In `lambda.ts`, we only export the `handler` function.
- In `main.ts`, we just call the `listen` method.

To test locally, we run `npm run start:debug` by calling `main.ts`. To deploy to AWS Lambda, we use `lambda.ts`.
It's not possible to use both because things can get in the way because both options start an http server when running.

Sorry for being redundant, English is not my native language so I'm trying my best to explain, also if you see any English errors in the documentation, please let me know and I'll fix it.

Reply:
I think the question was - what are the advantages of using `nest start` instead of `serverless offline` (and so maintaining 2 separate files)?

Reply:
With `nest start:debug` every time you change some file it will be reloaded. With `serverless offline`, your changes will only be reflected after you stop the command, run `nest build`, and then run `serverless offline` again.
As bchilcott said, it is unrealistic to have the old workflow as the main default.